proposed/MPID Messaging.md

```diff
@@ -0,0 +1,295 @@
+
+A large abuse of modern day digital communications is the ability of spammers and less than ethical marketing companies to flood the Internet with levels of unsolicited email to a level of circa 90%.  This is a waste of bandwidth and a significant nuisance.
+
+## Supported Use-Cases
+
+The system will support secure, private, bi-directional communications between pairs of Clients.
+
+## Expected Outcome
+
+The provision of a secure messaging system which will eradicate unwanted and intrusive communications.
+
+# Detailed design
+
+## Overview
+
+A fundamental principle is that the cost of messaging is with the sender, primarily to deter unsolicited messages.  To achieve this, the sender will maintain messages in a network [outbox][3] until they are retrieved by the recipient.  The recipient will have a network [inbox][4] comprising a list of metadata relating to messages which are trying to be delivered to it.
+
+If the message is unwanted, the recipient simply does not retrieve the message.  The sender will thus quickly fill their own outbox with undelivered mail and be forced to clean this up themselves before being able to send further messages.
+
+This paradigm shift will mean that the obligation to unsubscribe from mailing lists, etc. is now with the owner of these lists.  If people are not picking up messages, it is because they do not want them.  So the sender has to do a better job.  It is assumed this shift of responsibilities will lead to a better-managed bandwidth solution and considerably less stress on the network and the users of the network.
+
+## Implementation Details
+
+The two relevant structs (other than the MPID itself which is a [standard Client key][0]) are the [`MpidHeader`][1] and the [`MpidMessage`][2].
+
+Broadly speaking, the `MpidHeader` contains metadata and the `MpidMessage` contains a signed header and message.  We also want to define some upper limits which will be described later.
+
+### Consts
+
+```rust
+pub const MPID_MESSAGE_TAG: u64 = 51000;
+pub const MPID_HEADER_TAG: u64 = 51001;
+pub const MAX_SUBJECT_SIZE: usize = 128;  // bytes
+pub const MAX_INBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+pub const MAX_OUTBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+```
+
+### `MpidHeader`
+
+```rust
+pub struct MpidHeader {
+    pub sender: ::routing::NameType,
+    pub recipient: ::routing::NameType,
+    pub index: u32,
+    pub parent_index: Option<u32>,
+    pub subject: Vec<u8>,
+}
+```
+
+The `sender` and `recipient` fields are hopefully self-explanatory.  However, the only purpose for them existing in the `MpidHeader` is to increase the randomness of the derived message name so conflicts have a much lower chance of happening.
+
+The `index` and `parent_index` are intended to allow threading or sequencing of messages, while `subject` allows passing arbitrary data (e.g. a subject line).
```

Comment:
I agree with Qi that these would be better being replaced by a single timestamp field.  However, I also take your point Spandan that we don't explain this well enough.

If these fields remain in the submitted RFC, I'll add something like the following:

`index` and `parent_index` can be set to any arbitrary values the sender wishes - they are not used by the network other than to help disambiguate messages.  From the client's perspective, the intent is that they can be used to help create "threaded" conversations between users, i.e. a message can be the first of a new thread by setting `parent_index` to `None`, or it can be a reply to message with `index` x by setting `parent_index` to `Some(x)`.